packages/jsii/lib/assembler.ts

```diff
@@ -0,0 +1,879 @@
+import colors = require('colors/safe');
+import crypto = require('crypto');
+import deepEqual = require('deep-equal');
+import fs = require('fs-extra');
+import spec = require('jsii-spec');
+import log4js = require('log4js');
+import path = require('path');
+import ts = require('typescript');
+import { Diagnostic, Emitter } from './emitter';
+import literate = require('./literate');
+import { ProjectInfo } from './project-info';
+import utils = require('./utils');
+import { Validator } from './validator';
+
+// tslint:disable:no-var-requires Modules without TypeScript definitions
+const sortJson = require('sort-json');
+// tslint:enable:no-var-requires
+
+const LOG = log4js.getLogger('jsii/assembler');
+
+/**
+ * The JSII Assembler consumes a ``ts.Program`` instance and emits a JSII assembly.
+ */
+export class Assembler implements Emitter {
+    private _diagnostics = new Array<Diagnostic>();
+    private _deferred = new Array<() => void>();
+    private _types: { [fqn: string]: spec.Type };
+
+    /**
+     * @param projectInfo information about the package being assembled
+     * @param program     the TypeScript program to be assembled from
+     */
+    public constructor(public readonly projectInfo: ProjectInfo,
+                       public readonly program: ts.Program) {}
+
+    private get _typeChecker(): ts.TypeChecker {
+        return this.program.getTypeChecker();
+    }
+
+    /**
+     * Attempt emitting the JSII assembly for the program.
+     *
+     * @return the result of the assembly emission.
+     */
+    public async emit(): Promise<EmitResult> {
+        this._diagnostics = [];
+        if (!this.projectInfo.description) {
+            this._diagnostic(null,
+                             ts.DiagnosticCategory.Suggestion,
+                             'A "description" field should be specified in "package.json"');
+        }
+        if (!this.projectInfo.homepage) {
+            this._diagnostic(null,
+                             ts.DiagnosticCategory.Suggestion,
+                             'A "homepage" field should be specified in "package.json"');
+        }
+        const readme = await _loadReadme.call(this);
+        if (!readme) {
+            this._diagnostic(null,
+                             ts.DiagnosticCategory.Suggestion,
+                            'There is not "README.md" file. It is recommended to have one.');
+        }
+
+        this._types = {};
+        this._deferred = [];
+        const mainFile = path.resolve(this.projectInfo.projectRoot, this.projectInfo.types.replace(/\.d\.ts(x?)$/, '.ts$1'));
+        for (const sourceFile of this.program.getSourceFiles().filter(f => !f.isDeclarationFile)) {
+            if (sourceFile.fileName !== mainFile) { continue; }
+            if (LOG.isTraceEnabled()) {
```

Comment:
Why do you need to check if trace is enabled? Isn't that just a log level?

Reply:
It's a log level. The check is to avoid taking on possibly expensive side-effects (this string interpolates with color & all).

Reply:
Indeed sounds very expensive. Sometimes those loggers allow you to pass in a function which is lazy evaluated only if the level is enabled...

Reply:
Yeah but that actually makes the syntax of logging heavier. That's a bit annoying.

Reply:
wrapping every trace call with an "if" statement is also heavy...